beacon-chain/db/slasherkv/pruning.go

```diff
@@ -0,0 +1,157 @@
+
+	// We retrieve the lowest stored slot in the proposals bucket.
+	var lowestSlot types.Slot
+	if err = s.db.View(func(tx *bolt.Tx) error {
+		proposalBkt := tx.Bucket(proposalRecordsBucket)
+		c := proposalBkt.Cursor()
+		k, _ := c.First()
+		lowestSlot = slotFromProposalKey(k)
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// If the lowest slot is greater than or equal to the end pruning slot,
+	// there is nothing to prune, so we return early.
+	if lowestSlot >= endPruneSlot {
+		return nil
+	}
+
+	// We prune in increments of `pruningEpochIncrements` at a time to prevent
+	// a long-running bolt transaction which overwhelms CPU and memory.
+	slotCursor := lowestSlot
+	var encodedSlotCursor []byte
+
+	// While we still have epochs to prune based on a cursor, we continue the pruning process.
+	for epochAtCursor := helpers.SlotToEpoch(slotCursor); endEpoch-epochAtCursor > 0; {
+
+		// Each pruning iteration involves a unique bolt transaction. Given pruning can be
+		// a very expensive process which puts pressure on the database, we perform
+		// the process in a batch-based method using a cursor to proceed to the next batch.
+		log.Debugf("Pruned %d/%d epochs worth of proposals", endEpoch-epochAtCursor, endEpoch)
+		encodedSlotCursor = fssz.MarshalUint64([]byte{}, uint64(slotCursor))
+		if err = s.db.Update(func(tx *bolt.Tx) error {
+			proposalBkt := tx.Bucket(proposalRecordsBucket)
+			c := proposalBkt.Cursor()
+
+			var lastPrunedEpoch, epochsPruned types.Epoch
+			// We begin a pruning iteration at starting from the current slot cursor.
+			for k, _ := c.Seek(encodedSlotCursor); k != nil; k, _ = c.Next() {
+				// We check the slot from the current key in the database.
+				// If we have hit a slot that is greater than the end slot of the pruning process,
+				// we then completely exit the process as we are done.
+				if !slotPrefixLessThan(k, encodedEndSlot) {
+					// We don't want to unmarshal the key bytes every time
+					// into the cursor value, so we only do it if needed here.
+					slotCursor = slotFromProposalKey(k)
+					return nil
+				}
+				slot := slotFromProposalKey(k)
+				epoch := helpers.SlotToEpoch(slot)
+				slotCursor = slot
+
+				// If we have pruned N epochs in this pruning iteration,
+				// we exit from the bolt transaction.
+				if epochsPruned >= pruningEpochIncrements {
+					return nil
+				}
+
+				// Proposals in the database look like this:
+				//  (slot ++ validatorIndex) => encode(proposal)
+				// so it is possible we have a few adjacent objects that have the same slot, such as
+				//  (slot = 3 ++ validatorIndex = 0) => ...
+				//  (slot = 3 ++ validatorIndex = 1) => ...
+				//  (slot = 3 ++ validatorIndex = 2) => ...
+				// so we only mark an epoch as pruned if the epoch of the current object
+				// under the cursor has changed.
+				if epoch != lastPrunedEpoch {
+					epochsPruned++
+					lastPrunedEpoch = epoch
+				}
```

Comment:
Just to double-check (being paranoid here): this will only work fine if we are guaranteed that slots are coming in sorted order, if keys happen to come out of order than epochs are not strictly non-decreasing, and we might count more epochs pruned as we actually did (ex: if slots come as 31, 32, 30, 33 then we will have epochs 0, 1, 0, 1 -- which will count 3 epochs pruned, while we are crunching through only four slots of two different epochs).

Reply:
Excellent feedback, we need to check that the epoch has grown instead of just changed